feat: added code to deal with self signed certs

z103cb opened the pull request:
Added self signed certs and fixed the grpc / http clients to support the ssl protocol using self signed certificates

Fixes #8

z103cb committed:
feat: :bug: added code to deal with self signed certs

Added self signed certs and fixed the grpc / http clients to support the ssl protocol using self signed certificates

Fixes #8

dtrifiro committed:
tests: parametrize tls insecure

dtrifiro committed:
tests: cleanup caikit_nlp_runtime config setup

dtrifiro committed:
tests: use randomized ports for http/grpc servers

dtrifiro committed:
http client: add post init to HTTPConfig

we must validate exclusive tls/mtls options

dtrifiro committed:
http client: add asserts to satisfy mypy

z103cb committed:
http client: additional asserts for mypy

z103cb committed:
tests: reformating

z103cb committed:
[tests] disabled insecure tests

z103cb committed:
tests: post rebase fixes

z103cb committed:
[tests] fixed linter error

z103cb committed:
[tests] fixed missing fixture error

dtrifiro commented on tests/conftest.py:
```diff
@@ -13,6 +14,22 @@
+
+WAIT_TIME_OUT = 10
+
+CA_CERT_FILE = str(Path(__file__).parent / "resources/ca.pem")
+CLIENT_KEY_FILE = str(Path(__file__).parent / "resources/client-key.pem")
+CLIENT_CERT_FILE = str(Path(__file__).parent / "resources/client.pem")
+SERVER_KEY_FILE = str(Path(__file__).parent / "resources/server-key.pem")
+SERVER_CERT_FILE = str(Path(__file__).parent / "resources/server.pem")
```
These seem duplicated

z103cb replied:
will fix.

z103cb replied:
Fixed in [ee02430](https://github.com/opendatahub-io/caikit-nlp-client/pull/16/commits/ee02430be42d16ef76793be70ee9616d23268997)

z103cb committed:
fix: removed duplicate constants

Removed the duplicate definition of cert files and wait constants

dtrifiro commented on src/caikit_nlp_client/http_client.py:
```diff
@@ -95,9 +115,22 @@ def generate_text_stream(self, model_id: str, text: str, **kwargs) -> list[str]:
-            response = requests.post(  # FIXME: just a stub implementation
-                self.stream_api_url, json=json_input, timeout=10
-            )
+            if self.config.tls:
+                assert self.config.client_key_path
+                assert self.config.client_crt_path
```
Do we need the `client_key_path` and `client_crt_path` for `tls`?  Don't we need these two for `mTLS` only

dtrifiro replied:
See the suggestion below

dtrifiro commented on src/caikit_nlp_client/http_client.py:
```diff
@@ -61,7 +69,19 @@ def generate_text(self, model_id: str, text: str, **kwargs) -> str:
             log.info(f"Calling generate_text for '{model_id}'")
             json_input = create_json_request(model_id, text, **kwargs)
-            response = requests.post(self.api_url, json=json_input, timeout=10)
+            if self.config.tls:
+                assert self.config.client_key_path
+                assert self.config.client_crt_path
+
+                response = requests.post(
+                    self.api_url,
+                    json=json_input,
+                    timeout=10.0,
+                    verify=self.config.ca_crt_path,
+                    cert=(self.config.client_crt_path, self.config.client_key_path),
+                )
+            else:
+                response = requests.post(self.api_url, json=json_input, timeout=10.0)
```
```suggestion
            kwargs = {}
            
            if self.config.tls or self.config.mtls:
                kwargs['verify'] = self.config.ca_crt_path
            if self.config.mtls:
                 kwargs['cert'] = (self.config.client_crt_path, self.config.client_key_ath)

            response = requests.post(
                self.api_url,
                json=json_input,
                timeout=10.0,
                **kwargs
            )
```

z103cb replied:
Will do

dtrifiro commented on tests/conftest.py:
```diff
@@ -47,8 +55,15 @@ def model_name():
+# connections
 @pytest.fixture(scope="session")
-def caikit_nlp_runtime(grpc_server_port, http_server_port):
+def insecure() -> bool:
```
Waiting for the parametrized fixture

dtrifiro commented on tests/conftest.py:
```diff
@@ -76,72 +91,92 @@ def caikit_nlp_runtime(grpc_server_port, http_server_port):
-        port=port,
-        insecure=True,  # TODO: handle cases with secure=True with self signed certs
-    )
+def channel_factory(host: str, port: int, insecure: bool):
```
This will probably have to be parametrized with both tls=true and mtls=true. We'll do this in a separate PR

z103cb committed:
fix: skipping streaming tests and parametrization

Added skips for the streaming tests that are hanging even if successful.
Added parametrization over secure, insecure.
